app/jobs/service_history/rebuild_enrollments_job.rb

```diff
@@ -33,14 +33,22 @@ def perform
 
         # If this client has been invalidated, remove all service history and rebuild
         client.force_full_service_history_rebuild if client.service_history_invalidated?
-        # You must join in the project here or it will try to rebuild enrollments
-        # with no project
-        enrollments = GrdaWarehouse::Hud::Client.where(id: client_id).
-          joins(source_enrollments: :project).
-          pluck(Arel.sql(e_t[:id].as('enrollment_id').to_sql))
-        Rails.logger.info "===RebuildEnrollmentsJob=== Processing #{enrollments.size} enrollments for #{client_id}"
+
+        enrollment_ids = []
+        GrdaWarehouse::Hud::Enrollment.with_advisory_lock('rebuild_enrollments') do
+          # You must join in the project here or it will try to rebuild enrollments
+          # with no project
+          enrollment_ids = GrdaWarehouse::Hud::Client.where(id: client_id).
+            joins(source_enrollments: :project).
+            merge(GrdaWarehouse::Hud::Enrollment.unassigned).
+            pluck(Arel.sql(e_t[:id].as('enrollment_id').to_sql))
```

Comment:
This should also find any enrollment without a corresponding SHE.entry record.  Then we can probably use it to replace `GrdaWarehouse::Tasks::ServiceHistory::Add`

Reply:
```
 enrollment_ids = GrdaWarehouse::Hud::Client.where(id: client_id).
             joins(source_enrollments: :project).
             merge(GrdaWarehouse::Hud::Enrollment.unassigned).
             pluck(Arel.sql(e_t[:id].as('enrollment_id').to_sql))

enrollment_ids += GrdaWarehouse::Hud::Enrollment.joins(:project, :destination_client).where.not(id: GrdaWarehouse::ServiceHistoryEnrollment.entry.joins(:enrollment).select(e_t[:id])).pluck(:id)

enrollment_ids += GrdaWarehouse::Hud::Client.without_service_history.joins(:source_enrollments).pluck(e_t[:id])

enrollment_ids.uniq!
```

Reply:
I reworked the enrollment processing, but did not remove `Add` etc... as they seemed a useful abstraction.